---
title: "stCancer"
author: "wguo"
date: "2021/1/1"
output: html_document
---

<style type="text/css">
    body{
        font-size: 15px;
        line-height: 22px;
    }
    h1.title {
        font-size: 38px;
    }
    h1 {
        font-size: 28px;
        margin-top: 23px;
    }
    h2 {
        font-size: 24px;
        margin-top: 25px;
    }
    h3 {
      font-size: 20px;
        margin-top: 25px;
    }
    code.r{
        font-size: 13px;
    }
    pre {
        font-size: 14px;
    }
    p {
        margin-top:10px;
        margin-bottom:10px;
    }
    table { 
        width: 60%;
        border-collapse: collapse;
        font-family: Futura, Arial, sans-serif;
    }
    th,td {
        padding: 5px;
    }
    th,td {
        border-bottom: 1px solid #ddd;
        border-top: 1px solid #ddd;
        padding-right: 20px
    }
</style>


```{r setting, include=FALSE}
options(knitr.table.format = "html") 
options(scipen=10)
knitr::opts_chunk$set(echo = TRUE, fig.path = savePath_basic)
knitr::opts_knit$set(root.dir = savePath_basic)

title <- "stCancer"
if(!is.null(sampleName)){
    title <- paste0(sampleName, "  -  ", title)
}

if(!is.null(authorName)){
    userName <- authorName
}else{
    userName <- Sys.getenv("USERNAME")
}
reportMark <- Sys.time()
if(userName != ""){
    reportMark <- paste0(userName, " , ", reportMark)
}

h.i <- 1
h.ii <- 1
```

# `r title`
--------------------------------
<p align="right">`r reportMark`</p>


## `r h.i` Spot statistics

* The input of `stCancer` pipeline is the matrix generated by [Spatial Ranger](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger).

```{r echo=F, results='asis'}
if(file.exists(file.path(dataPath, "web_summary.html"))){
  file.copy(file.path(dataPath, "web_summary.html"), 
            file.path(savePath_basic, "report-spaceRanger.html"), 
            overwrite = T)
  cat("* Here is the [summary report](./report-spaceRanger.html) from `Space Ranger`.", sep = "")
}
```

### `r h.i`.`r h.ii` The number of UMIs and detected genes in spots

We display `nUMI` (total number of UMIs) and `nGene` (total number of detected genes) here.
 
```{r sp, echo=F, message=F, warning=F, dpi=300, fig.height=2.5, fig.width=8}
knitr::include_graphics(file.path(savePath_basic, "QC/basic.png"))
```
<p align='right' style='margin-top:3px'>(Hi-res image: <a href='./QC/basic.png')>view</a>)</p>


```{r echo=F}
h.i <- h.i + 1
h.ii <- 1
```






## `r h.i` Gene statistics
The number of genes expressed in at least one cell : ` `r sum(nSpot > 0)` `.


### `r h.i`.`r h.ii` Mitochondrial genes
Summary of mitochondrial genes percentage (`mito.percent`) in cells:
```{r mito.summary, echo=F}
format(summary(spatial@meta.data[["mito.percent"]]), digits = 3)
```

```{r ribo, echo=FALSE, message=F, dpi=300, fig.height=4, fig.width=4, fig.align="center", out.width='70%'}
knitr::include_graphics(file.path(savePath_basic, "QC/mito-npc.png"))
```
<p align='right' style='margin-top:3px'>(Hi-res image: <a href='./QC/mito-npc.png'>view</a>)</p>


```{r echo=F}
h.ii <- h.ii + 1
```


### `r h.i`.`r h.ii` Ribosome genes
Summary of ribosome genes percentage (`ribo.percent`) in cells:
```{r ribo.summary, echo=F}
format(summary(spatial@meta.data[["ribo.percent"]]), digits = 3)
```

```{r mito, echo=FALSE, message=F, dpi=300, fig.height=4, fig.width=4, fig.align="center", out.width='70%'}
knitr::include_graphics(file.path(savePath_basic, "QC/ribo-npc.png"))
```
<p align='right' style='margin-top:3px'>(Hi-res image: <a href='./QC/ribo-npc.png'>view</a>)</p>

```{r echo=F}
h.ii <- h.ii + 1
```



### `r h.i`.`r h.ii` Ambient RNAs 

### `r h.i`.`r h.ii`.1 Highly-expressed genes
In order to analyze the gene expression profiles in detail and identify 
highly-expressed genes in background mRNAs from lysed cells, 
we calculate some metrics as shown below.

* `prop.median`	: the median of expression proportions for a gene in each tissue spot.
* `detect.rate`	: the detected (`#UMI > 0`) rate for a gene in all tissue spots.

Here is a plot showing the distributions of gene proportion in spots for the first 100 genes. And the points (genes) are colored according to whether they belongs to mitochondrial, ribosome, or other genes.

```{r genePropPlot, echo=F, message=F, warning=F, dpi=300, fig.width=8, fig.height=8, fig.align="center"}
knitr::include_graphics(file.path(savePath_basic, "QC/geneProp.png"))
```
<p align="right">(Hi-res image: <a href='./QC/geneProp.png'>view</a>)</p>

The plot below shows the relationship between `detect.rate` and `prop.median`.

```{r geneRel, echo=F, message=F, warning=F, dpi=300, fig.width=4, fig.height=4, fig.align="center", out.width='40%'}
knitr::include_graphics(file.path(savePath_basic, "QC/gene-dr-med.png"))
```
<p align="right">(Hi-res image: <a href='./QC/gene-dr-med.png'>view</a>)</p>



```{r echo=F}
h.i <- h.i + 1
h.ii <- 1
```


## `r h.i` Output

### `r h.i`.`r h.ii` Threshold to filter spots
In the process above, we use ` `r paste0("region.threshold : ", region.threshold)` ` to filter the isolated spots, and ` `r ncol(object)` ` spots are left.

```{r echo=F}
h.ii <- h.ii + 1
```

### `r h.i`.`r h.ii` Output files
Running this script generates following files:

1. **Html report** :
[report-stStat.html](./report-stStat.html).
2. **Markdown report** :
[report-stStat.md](./report-stStat.md).
3. **Figure files** :
[QC/](./QC/).
4. **Data**:
[data/](./data/).
5. **Text file with gene manifest** :
[geneManifest.txt](./data/geneManifest.txt).
6. **SeuratObject after stStatistic** :
[SeuratObject](./data/spatial_stat.RDS).
